Add testing task 3

tsbonev opened the pull request:
Created Storage class that can keep track of products and their quantities in a HashMap
Created custom exceptions to be thrown when a product tries to go above max quantity, 0 quantity or does not exist in the map
Created a comparator that sorts the keyset of the map by price after they have been parsed into a List
Wrote tests that check for proper logic execution, invalid arguments and exception throws

mgenov commented on src/com/clouway/crm/core/Storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+
+public class Storage {
+
+    private Map<Product, Integer> items;
```
s/items/inventory as it's the inventory of the Store.

mgenov commented on src/com/clouway/crm/core/Storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+     * @throws NoSuchProductException exception if the product does not exist
+     * @throws MaxQuantityException exception if the product does not allow the sought quantity increase
+     */
+    public void stock(String name, int amount)
```
adjustStockLevel or addStock or  will be more meaningful

mgenov commented on src/com/clouway/crm/core/Storage/Storage.java:
```diff
@@ -0,0 +1,132 @@
+     * Sorts the keys of the map by price in descending order.
+     * @return a sorted list of the keys
+     */
+    public List sort(){
```
A generic list should be returned as it will indicate that this list contains Product objects. Also the name of the method could be improved as this is what the description of the task says:

"Да се добави възможността продуктите в склада да се извеждат сортирани по цена."

e.g getProducts() may return them by price.

mgenov commented on tests/com/clouway/crm/core/Storage/StorageTest.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    @Test
+    void genericObjectInAdd() {
```
This test doesn't look reasonable as type-safety of java is ensuring this stuff.

mgenov commented on tests/com/clouway/crm/core/Storage/StorageTest.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    @Test
+    void nullPointerInAdd(){
```
addNullableProduct

In kotlin this is ensured with the type definition. If you define your method/function like:
```kotlin
fun add(product: Product)
```

then the compiler will fail with error if you try to pass null 

Otherwise if function is defined as:

```kotlin
fun add(product: Product?)
```

e.g what I'm trying to say is that Kotlin provides type improvements.

mgenov commented on tests/com/clouway/crm/core/Storage/StorageTest.java:
```diff
@@ -0,0 +1,114 @@
+        sortedList.add(storage.getProduct("white bread"));
+        sortedList.add(storage.getProduct("bread"));
+
+        assertTrue(sortedList.equals(storage.sort()));
```
Hamcrest has matchers for this:

```java
assertThat(storage.sort(), is(equalTo(Lists.newArrayList(
     new Product("luxury bread"),
     new Product("white bread"),
     new Product("bread"),
));
```

storage.getProduct looks like is exposed only for tests which is not a good idea.

mgenov commented on tests/com/clouway/crm/core/Storage/StorageTest.java:
```diff
@@ -0,0 +1,114 @@
+    void setup(){
+        storage = new Storage();
+
+        Product bread = new Product("bread", 6,1);
```
Don't put test related logic in setup unless it's required due some reason. When I read the test removeProduct I see storage.remove("white bread") but I'm not able to see where is added as adding is defined in the setup which makes test harder to read.

tsbonev commented:
Changed test names to better show their use
Added hamcrest and guava, refactored assert statements to use hamcrest
Made getProduct private and changed the sort test to not rely on it
Changed method names in Storage to better match their function
Renamed items list to sInventory
Changed return on the sort method to return a generic List<Product>

mgenov commented on src/com/clouway/crm/core/storage/Storage.java:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * Comparator that sorts by price in descending order.
+     */
+    private class PriceComparator implements Comparator<Product>{
```
private methods should be defined last in the class. 

https://github.com/clouway/teamspace/wiki/Java-Code-Convention

mgenov commented on src/com/clouway/crm/core/storage/Storage.java:
```diff
@@ -0,0 +1,130 @@
+     *
+     * @return the size of the valuesets in a map
+     */
+    public int getSize(){
```
Doesn't sound reasonable size to be exposed to the callers.

mgenov commented on src/com/clouway/crm/core/storage/Storage.java:
```diff
@@ -0,0 +1,130 @@
+     * @param name of the product that is sought.
+     * @return the product or null if it is not found
+     */
+    private Product getProduct(String name){
```
getProduct should not be exposed to the callers from my point of view. 

Here is a sample snipped which describes some testing behaviour: 
```java
store.addStock("Apples", 10);
store.sell("Apples", 5);
List<StockItem> stockItems = store.availableStock();
assertThat(stockItems, is(equalTo(Lists.newArrayList(StockItem("Apples", 5)));
```

mgenov replied:
When store purchase stock items (the store buys stock to be able to sell it)  it's with SKU(Name), Quantity and Buy Price. When you sell something, the client provides name and quantity and the store determines the price of the product which to be the sell price. The goal of my comment was to point you out that these are 2 different things and you are exposing this method as one object is used for two things. 

If we follow this flow, maybe the design becomes like: 

```java
store.purchaseStock("Apples", 10, 2.30)
Order order = stock.createSellOrder("Apples", 5)
// order has sell price of the Applies
// accept order
Invoice invoice = store.payOrder(order) 
// stock.cancelOrder(order) User don't want to pay as the price is too big
```

tsbonev replied:
I can change the addStock method to accept fields and construct its own Product instance which it will then add as a key to the map, but I still don't understand where getProduct can be replaced, as it is used when removing stock which only takes a name, selling stock and increasing stock which take a name and amount but not adding new stock completely.

mgenov replied:
Just saw that it's private. SGTM in this case, so you can keep the current design.

mgenov commented:
Please squash your commits, to be able to merge it.

tsbonev commented:
How far back should I squash the commits?

mgenov commented:
1 PR = 1 commit

tsbonev commented:
Okay, good to know.
